syntax = "proto3";

package aiserver.v1;

message StreamCppRequest {
  CurrentFileInfo current_file = 1;
  repeated string diff_history = 2;
  optional string model_name = 3;
  optional LinterErrors linter_errors = 4;
  repeated string diff_history_keys = 5;
  optional bool give_debug_output = 6;
  repeated FileDiffHistory file_diff_histories = 7;
  repeated FileDiffHistory merged_diff_histories = 8;
  repeated BlockDiffPatch block_diff_patches = 9;
  optional bool is_nightly = 10;
  optional bool is_debug = 11;
  optional bool immediately_ack = 12;
  repeated ContextItem context_items = 13;
  repeated ParameterHint parameter_hints = 14;
  repeated LspContext lsp_contexts = 15;
  optional CppIntentInfo cpp_intent_info = 16;
  optional bool enable_more_context = 17;
  optional string workspace_id = 18;
  repeated AdditionalFile additional_files = 19;
  optional ControlToken control_token = 20;
  optional double client_time = 21;
  repeated FilesyncUpdate filesync_updates = 22;
  double time_since_request_start = 23;
  double time_at_request_send = 24;
  optional double client_timezone_offset = 25;
  optional LspSuggestedItems lsp_suggested_items = 26;
  optional bool supports_cpt = 27;
  optional bool supports_crlf_cpt = 28;
  repeated CodeResult code_results = 29;
}

message CurrentFileInfo {
  string relative_workspace_path = 1;
  string contents = 2;
  CursorPosition cursor_position = 3;
  repeated DataFrame dataframes = 4;
  string language_id = 5;
  Selection selection = 6;
  repeated Diagnostic diagnostics = 7;
  int32 total_number_of_lines = 8;
  int32 contents_start_at_line = 9;
  repeated TopChunk top_chunks = 10;
  optional int32 alternative_version_id = 11;
  optional int32 file_version = 14;
  repeated int32 cell_start_lines = 15;
  repeated Cell cells = 16;
  optional string sha_256_hash = 17;
  bool rely_on_filesync = 18;
  string workspace_root_path = 19;
  optional string line_ending = 20;
}

message DataFrame {}
message Diagnostic {}
message TopChunk {}
message Cell {}

message CursorPosition {
  int32 line = 1;
  int32 column = 2;
}

message LinterErrors {
  string relative_workspace_path = 1;
  repeated LinterError errors = 2;
  string file_contents = 3;
}

message LinterError {
  string message = 1;
  Selection range = 2;
  optional string source = 3;
  repeated RelatedInformation related_information = 4;
  optional Severity severity = 5;
  optional bool is_stale = 6;
}

message Selection {
  int32 start_line = 1;
  int32 start_column = 2;
  int32 end_line = 3;
  int32 end_column = 4;
}

message RelatedInformation {
  string message = 1;
  string relative_workspace_path = 2;
  repeated string relevant_lines = 3;
  int32 start_line = 4;
}

message FileDiffHistory {
  string file_name = 1;
  repeated string diff_history = 2;
  repeated double diff_history_timestamps = 3;
}

message BlockDiffPatch {}

message ContextItem {}

message ParameterHint {}

message LspContext {}

message CppIntentInfo {
  string source = 1;
}

message AdditionalFile {
  string relative_workspace_path = 1;
  bool is_open = 2;
  repeated string visible_range_content = 3;
  optional double last_viewed_at = 4;
  repeated int32 start_line_number_one_indexed = 5;
  repeated LineRange visible_ranges = 6;
}

message LineRange {
  int32 start_line_number = 1;
  int32 end_line_number_inclusive = 2;
}

message FilesyncUpdate {}

message LspSuggestedItems {
  repeated LspSuggestion suggestions = 1;
}

message LspSuggestion {}

message CodeResult {
  CodeBlock code_block = 1;
  double score = 2;
}

message CodeBlock {
  string relative_workspace_path = 1;
  Range range = 2;
  string contents = 3;
}

message Range {
  Position start_position = 1;
  Position end_position = 2;
}

message Position {
  int32 line = 1;
  int32 column = 2;
}

enum ControlToken {
  CONTROL_TOKEN_UNSPECIFIED = 0;
  CONTROL_TOKEN_QUIET = 1;
  CONTROL_TOKEN_LOUD = 2;
  CONTROL_TOKEN_OP = 3;
}

enum Severity {
  SEVERITY_UNSPECIFIED = 0;
}
